src/ast/nodes/shared/MultiExpression.ts

```diff
@@ -0,0 +1,65 @@
+import CallOptions from '../../CallOptions';
+import { DeoptimizableEntity } from '../../DeoptimizableEntity';
+import { ExecutionPathOptions } from '../../ExecutionPathOptions';
+import { ImmutableEntityPathTracker } from '../../utils/ImmutableEntityPathTracker';
+import { LiteralValueOrUnknown, ObjectPath, UNKNOWN_VALUE } from '../../values';
+import { ExpressionEntity } from './Expression';
+
+export class MultiExpression implements ExpressionEntity {
```

Comment:
This is a big step. Intermediate expressions!

Could function returns and generator yields not be treated as MultiExpression of all their return values?

Reply:
The problem with function returns turned out to be a problem of convergence order for some code-bases (say bundling the TypeScript compiler for instance). If there are many functions with many returns, then with only a few steps, a single `hasEffects` check could suddenly trigger a million secondary checks as all returnExpressions are checked. This should definitely be revisited, though. One might consider e.g. limiting the number of tracked return expressions to two or similar things.

Reply:
Definitely, setting a limit on this would be a good starting approach!